lib/rules/no-mutation-props.js

```diff
@@ -0,0 +1,193 @@
+      if (topObject.object && topObject.object.type === 'ThisExpression') {
+        return topObject.property && topObject.property.name === 'props';
+      }
+
+      // if this is a variable, we need to look at the scope to determine if
+      // it is actually a reference to a prop
+      const name = topObject.name || (topObject.object && topObject.object.name);
+      const originalDeclaration = context.getScope().variableScope.set.get(name);
+      if (originalDeclaration) {
+        const originalIdentifiers = originalDeclaration.identifiers;
+        // if the get call returns falsey, we've not set it in our cache
+        // and it can't have have been from props. See `VariableDeclaration`
+        const currentIdentifier = components.get(originalIdentifiers[originalIdentifiers.length - 1]);
+        return currentIdentifier && currentIdentifier.isFromProps;
+      }
+
+      return false;
+    }
+
+    function setPropMutation(mutation, node) {
+      const component = components.get(utils.getParentComponent());
+      const propMutations = component && component.propMutations
+        ? component.propMutations.concat([mutation])
+        : [mutation];
+      components.set(node || mutation, {propMutations});
+    }
+
+    /**
+     * Reports this.props mutations for a given component
+     * @param {Object} component The component to process
+     */
+    function reportMutations(component) {
+      if (!component.propMutations) {
+        return;
+      }
+
+      component.propMutations.forEach(mutation => {
+        context.report({
+          node: mutation,
+          message: 'A component must never modify its own props.'
+        });
+      });
+    }
+
+    // --------------------------------------------------------------------------
+    // Public
+    // --------------------------------------------------------------------------
+
+    return {
+      VariableDeclaration: function(node) {
+        if (!node.declarations) {
+          return;
+        }
+
+        node.declarations.forEach(declaration => {
+          const init = declaration.init;
+          const id = declaration.id;
+          const isProps = isPropsNode(init);
+          if (!isProps) {
+            return;
+          }
+          const assignments =
+            // if this is a simple assignment `const b = this.props.b`
+            (id.type === 'Identifier' && [id])
+            // if this is object destructuring `const {b} = this.props`
+            || (id.type === 'ObjectPattern' && id.properties.map(property => property.value))
+            // if this is array destructuring `const [b] = this.props.list`
+            || (id.type === 'ArrayPattern' && id.elements);
+
+          // remember that this node is a prop, `isFromProps` relies on this
+          assignments.forEach(assignment => {
+            components.add(assignment);
+            components.set(assignment, {
+              isFromProps: true
```

Comment:
I'm not sure how I feel about using the `components` as a storage for non-component nodes... Maybe we can use some local collection instead?

Reply:
Fair enough, I was just trying to re-use the logic since what I need is all there. But, I'm happy to copy/paste what I need into here?

Reply:
Something like:

```
          assignments.forEach(assignment => {
            allAssignments.push(assignment.range.join(':'));
          });
```

And where it's checked:

```
        const currentIdentifier = allAssignments.indexOf(originalIdentifiers[originalIdentifiers.length - 1].range.join(':')) >= 0;
        return currentIdentifier;
```

This way, at the end of the file there will only be the actual components inside the list, and we skip looping over all those assignments. For all tests, list would only contain 1 value.

```
        Object.keys(list).forEach(key => {
          const component = list[key];

          reportMutations(component);
        });
```

What do you think?